src/value.rs

```diff
@@ -300,6 +300,76 @@ impl Value {
             n,
         })
     }
+
+    /// Prune the value down to the given type.
+    ///
+    /// The pruned type must be _smaller than or equal to_ the current type of the value.
+    /// Otherwise, this method returns `None`.
+    ///
+    /// ## Smallness
+    ///
+    /// - `T` ≤ `T` for all types `T`
+    /// - `1` ≤ `T` for all types `T`
+    /// - `A1 + B1` ≤ `A2 + B2` if `A1` ≤ `A2` and `B1` ≤ `B2`
+    /// - `A1 × B1` ≤ `A2 × B2` if `A1` ≤ `A2` and `B1` ≤ `B2`
+    ///
+    /// ## Pruning
+    ///
+    /// - `prune( v: T, 1 )` = `(): 1`
+    /// - `prune( L(l): A1 + B1, A2 + B2 )` = `prune(l: A1, A2) : A2 + B2`
+    /// - `prune( R(r): A1 + B1, A2 + B2 )` = `prune(r: B1, B2) : A2 + B2`
+    /// - `prune( (l, r): A1 × B1, A2 × B2 )` = `( prune(l: A1, A2), prune(r: B1, B2): A2 × B2`
+    pub fn prune(&self, pruned_ty: &Final) -> Option<Self> {
+        enum Task<'a> {
+            Prune(&'a Value, &'a Final),
+            MakeLeft(Arc<Final>),
+            MakeRight(Arc<Final>),
+            MakeProduct,
+        }
+
+        let mut stack = vec![Task::Prune(self, pruned_ty)];
+        let mut output = vec![];
+
+        while let Some(task) = stack.pop() {
+            match task {
+                Task::Prune(value, pruned_ty) => match pruned_ty.bound() {
+                    CompleteBound::Unit => output.push(Value::unit()),
+                    CompleteBound::Sum(l_ty, r_ty) => {
+                        if let Some(l_value) = value.as_left() {
+                            stack.push(Task::MakeLeft(Arc::clone(r_ty)));
+                            stack.push(Task::Prune(l_value, l_ty));
+                        } else {
+                            let r_value = value.as_right()?;
+                            stack.push(Task::MakeRight(Arc::clone(l_ty)));
+                            stack.push(Task::Prune(r_value, r_ty));
+                        }
+                    }
+                    CompleteBound::Product(l_ty, r_ty) => {
+                        let (l_value, r_value) = value.as_product()?;
+                        stack.push(Task::MakeProduct);
+                        stack.push(Task::Prune(r_value, r_ty));
+                        stack.push(Task::Prune(l_value, l_ty));
+                    }
+                },
+                Task::MakeLeft(r_ty) => {
+                    let l_value = output.pop().unwrap();
+                    output.push(Value::left(l_value, r_ty));
+                }
+                Task::MakeRight(l_ty) => {
+                    let r_value = output.pop().unwrap();
+                    output.push(Value::right(l_ty, r_value));
+                }
+                Task::MakeProduct => {
+                    let r_value = output.pop().unwrap();
+                    let l_value = output.pop().unwrap();
+                    output.push(Value::product(l_value, r_value));
+                }
+            }
+        }
+
+        debug_assert_eq!(output.len(), 1);
+        output.pop()
```

Comment:
In 6436f6c221c37570cf37dc9557edb373a33019a6:

...and I would change this `debug_assert_eq` to an `assert_eq`. It's a super cheap check, we're only doing it once, and it'll catch some forms of "the passed type is not smaller than the original type".

Reply:
I think this assertion is guaranteed to be true for all inputs.

Reply:
I added unit tests for cases where the pruned type is not smaller than or equal to the value type.

Reply:
Yep, thanks! See above -- I missed the `?`s and though that there were no error returns up to this point.